conans/test/functional/toolchain/test_meson.py

```diff
@@ -0,0 +1,164 @@
+import textwrap
+import unittest
+from nose.plugins.attrib import attr
+
+from conans.client.tools.files import which
+from conans.model.version import Version
+from conans.test.utils.tools import TestClient
+from conans.util.files import decode_text
+from conans.util.runners import version_runner
+
+
+def get_meson_version():
+    try:
+        out = version_runner(["meson", "--version"])
+        version_line = decode_text(out).split('\n', 1)[0]
+        version_str = version_line.rsplit(' ', 1)[-1]
+        return Version(version_str)
+    except Exception as e:
+        return Version("0.0.0")
+
+
+@attr("slow")
+@attr("toolchain")
+@unittest.skipUnless(which("meson") and get_meson_version() >= Version("0.56.0"),
+                     "requires meson >= 0.56.0")
+class MesonToolchainTest(unittest.TestCase):
+    _conanfile_py = textwrap.dedent("""
+    from conans import ConanFile, MesonToolchain, tools
+
+
+    class App(ConanFile):
+        settings = "os", "arch", "compiler", "build_type"
+        options = {"shared": [True, False], "fPIC": [True, False]}
+        default_options = {"shared": False, "fPIC": True}
+
+        def config_options(self):
+            if self.settings.os == "Windows":
+                del self.options.fPIC
+
+        def toolchain(self):
+            tc = MesonToolchain(self)
+            tc.definitions["STRING_DEFINITION"] = "Text"
+            tc.definitions["TRUE_DEFINITION"] = True
+            tc.definitions["FALSE_DEFINITION"] = False
+            tc.definitions["INT_DEFINITION"] = 42
+            tc.definitions["ARRAY_DEFINITION"] = ["Text1", "Text2"]
+            tc.write_toolchain_files()
+
+        def build(self):
+            # this will be moved to build helper eventually
+            with tools.vcvars(self) if self.settings.compiler == "Visual Studio" else tools.no_op():
+                self.run("meson setup --native-file conan_meson_native.ini build .")
+                self.run("meson compile -C build")
+    """)
+
+    _meson_options_txt = textwrap.dedent("""
+    option('STRING_DEFINITION', type : 'string', description : 'a string option')
+    option('INT_DEFINITION', type : 'integer', description : 'an integer option', value: 0)
+    option('FALSE_DEFINITION', type : 'boolean', description : 'a boolean option (false)')
+    option('TRUE_DEFINITION', type : 'boolean', description : 'a boolean option (true)')
+    option('ARRAY_DEFINITION', type : 'array', description : 'an array option')
+    option('HELLO_MSG', type : 'string', description : 'message to print')
+    """)
+
+    _meson_build = textwrap.dedent("""
+    project('tutorial', 'cpp')
+    add_global_arguments('-DSTRING_DEFINITION="' + get_option('STRING_DEFINITION') + '"', language : 'cpp')
+    add_global_arguments('-DHELLO_MSG="' + get_option('HELLO_MSG') + '"', language : 'cpp')
+    hello = library('hello', 'hello.cpp')
+    executable('demo', 'main.cpp', link_with: hello)
+    """)
+
+    _hello_h = textwrap.dedent("""
```

Comment:
USe the new ``gen_function_cpp()`` and ``gen_function_h()`` and check later the underlying compiler, architecture used to build was the right one.

Reply:
used new functions `gen_functions_cpp`/`gen_functions_h`, also checked compiler and arch